sql: execution-time erroring of remote region access for enforce_home_region setting

msirek opened the pull request:
Informs #83819

enforce_home_region phase 2
---------------------------
This adds dynamic checking of whether locality-optimized searches
and locality-optimized lookup joins must access rows in remote regions.
If the `enforce_home_region` session setting is true, and remote access is
required in these operations, the query errors out. Both vectorized and
non-vectorized execution is handled.

Locality-optimized search is handled by setting an exclusive upper bound
on the inputs to the `SerialUnorderedSynchronizer` built for this operation,
and erroring out if the synchronizer iterates to that input before the query
finishes execution due to a LimitOp consumer.

Locality-optimized joins besides antijoin are handled by erroring out if
`joinReader.performLookup` calls `jr.strategy.generateRemoteSpans` and
is about to issue a lookup to a remote region.

Locality-optimized antijoins come as a pair, first the local lookup join,
and then the output of that feeds into a remote lookup join. This PR adds
tagging of that 2nd lookup join as having only lookups of remote rows so
that any attempted lookup by that 2nd join errors out.

Release note: None

physicalplan: pass ctx in more execution code paths
----

This commit passes the `context.Context` in all code paths necessary to
make it accessible to `PhysicalPlan.MergeResultStreams` and adds a `Ctx`
method to `exec.Factory` to make the context accessible in more locations.

Release note: None

Commit:
sql: execution-time erroring of remote region access for enforce_home_region setting

Informs #83819

enforce_home_region phase 2
---------------------------
This adds dynamic checking of whether locality-optimized searches
and locality-optimized lookup joins must access rows in remote regions.
If the `enforce_home_region` session setting is true, and remote access is
required in these operations, the query errors out. Both vectorized and
non-vectorized execution is handled.

Locality-optimized search is handled by setting an exclusive upper bound
on the inputs to the `SerialUnorderedSynchronizer` built for this operation,
and erroring out if the synchronizer iterates to that input before the query
finishes execution due to a LimitOp consumer.

Locality-optimized joins besides antijoin are handled by erroring out if
`joinReader.performLookup` calls `jr.strategy.generateRemoteSpans` and
is about to issue a lookup to a remote region.

Locality-optimized antijoins come as a pair, first the local lookup join,
and then the output of that feeds into a remote lookup join. This PR adds
tagging of that 2nd lookup join as having only lookups of remote rows so
that any attempted lookup by that 2nd join errors out.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/95484)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
Nice work! Overall, looks good to me (I didn't carefully think through the optimizer changes though), but I do have some suggestions and nits.

> Reviewed 23 of 23 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-:-NMAu1Rd9lPGQZinCxwp:bj13i9i)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msirek and @rytaft)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAAbX1BQzfSU9178gs):*
nit: we tend to prefix the commit message with a package (or packages) that are mostly affected by the commit, and I don't think we have `exec` package, perhaps `s/exec/sql/`? Also consider changing the commit title to something more understandable.
___
*[`pkg/sql/distsql_physical_planner.go` line 3968 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAtxiPDWJHmaPCy7NP:-NMAtxiPDWJHmaPCy7NQ:blmob2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/distsql_physical_planner.go#L3968)):*
> ```Go
> 			if n.enforceHomeRegion {
> 				serialStreamErrorSpec.Ctx = ctx
> 				serialStreamErrorSpec.SerialInputIdxExclusiveUpperBound = 1
> ```

nit: consider adding a quick comment for why we use `1` here.
___
*[`pkg/sql/union.go` line 100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAtpTwCwknpDjP3den:-NMAtpTwCwknpDjP3deo:bt2vpt) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/union.go#L100)):*
> ```Go
> 	// and the session setting `enforce_home_region` is true, indicating the
> 	// operation should error out if the query cannot be satisfied by accessing
> 	// only rows in the local region.
> ```

nit: perhaps mention that `left` input is only accessing local region whereas `right` is all about the remote regions.
___
*[`pkg/sql/execinfrapb/data.proto` line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMALGM63rgWLIP5-RX5:-NMALGM63rgWLIP5-RX6:bng7b08) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/execinfrapb/data.proto#L147)):*
> ```protobuf
> 
>   // EnforceHomeRegionError is the error to return when
>   // EnforceHomeRegionStreamExclusiveUpperBound is exceeded.
> ```

nit: should we explicitly mention that this error must be set if the upper bound is non-zero?
___
*[`pkg/sql/opt/xform/rules/join.opt` line 622 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAoud9C1BRwrCEOXrb:-NMAoudAAvGUUxXjTnap:bkassbb) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/opt/xform/rules/join.opt#L622)):*
> ```
>     )
>     $on
>     (CreateRemoteOnlyLookupJoinPrivate $remoteExpr $private)
> ```

nit: consider mentioning in the comment to this rule that handling of locality-optimized anti joins requires additional execution support, so this rule differs from other types of locality-optimized lookup joins.
___
*[`pkg/sql/physicalplan/physical_plan.go` line 173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMApGB529TBGvL0O1ng:-NMApGB529TBGvL0O1nh:b-dbyltk) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/physicalplan/physical_plan.go#L173)):*
> ```Go
> 
> 	// Ctx is the context for setting up the Error.
> 	Ctx context.Context
> ```

Storing the context here seems like an anti-pattern. We should instead pass `ctx` as an argument to `MergeResultStreams`. The plumbing there shouldn't be too bad.
___
*[`pkg/sql/rowexec/joinreader.go` line 344 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAq7iy4T2UCg2czxEN:-NMAq7iy4T2UCg2czxEO:b-qd8msg) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/rowexec/joinreader.go#L344)):*
> ```Go
> 	}
> 
> 	errorOnLookup := spec.RemoteOnlyLookups && flowCtx.EvalCtx.SessionData().EnforceHomeRegion &&
> ```

`EnforceHomeRegion` is part of local-only session data, meaning that it is not propagated to the remote nodes (if the plan is distributed). This shouldn't be an issue since plans with locality-optimized search full run on the gateway (i.e. they are local), but perhaps one day we'd ship it to a remote node to fully run. Should we make some assertions in the code about the plan being fully local? Probably it's not worth the effort, but I thought I'd mention this. WDYT?
___
*[`pkg/sql/rowexec/joinreader.go` line 1026 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAquxS7OSsVTxy_1tw:-NMAquxS7OSsVTxy_1tx:b1ue5nf) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/rowexec/joinreader.go#L1026)):*
> ```Go
> func (jr *joinReader) performLookup() (joinReaderState, *execinfrapb.ProducerMetadata) {
> 	if jr.errorOnLookup {
> 		err := pgerror.Newf(pgcode.QueryHasNoHomeRegion,
> ```

nit: we could extract this out to be a global singleton error that is also reused with the strategy.
___
*[`pkg/sql/rowflow/input_sync.go` line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMArHbz4QBexPmQmO4l:-NMArHbz4QBexPmQmO4m:b-xtf5me) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/rowflow/input_sync.go#L391)):*
> ```Go
> 	// the receiver when srcIndex has incremented to match
> 	// serialSrcIndexExclusiveUpperBound.
> 	exceedsSrcIndexExclusiveUpperBoundErrorFunc func()
> ```

Why is this a push function rather than an explicit error itself that would be returned on `serialUnorderedSynchronizer.Next` call?
___
*[`pkg/sql/rowflow/input_sync.go` line 409 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMArjzc18NSQLe-khha:-NMArjzc18NSQLe-khhb:bvwn31i) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/rowflow/input_sync.go#L409)):*
> ```Go
> 			u.srcIndex++
> 			if u.serialSrcIndexExclusiveUpperBound > 0 && u.srcIndex >= int(u.serialSrcIndexExclusiveUpperBound) &&
> 				(u.state == notInitialized || u.state == returningRows) {
> ```

nit: consider adding some commentary here why we don't want to error out with `draining` and `drainBuffered` states.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAs74q6JY_dPoNQgxt:-NMAs74r8wbMvrJcIwOm:b-uwi1rn) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L13)):*
> ```
> 
> statement ok
> SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false;
> ```

The auto stats collection should already be disabled (it is done in `logic.go:1675`). What prompted you to add this?
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 784 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAs_2b9HqHWlzL5U_q:-NMAs_2b9HqHWlzL5U_r:b-jiipme) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L784)):*
> ```
> 
> # ---------------------------
> # enforce_home_region phase 2
> ```

nit: consider linking the issue number that talks about phase 2.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 809 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAskVu4iU6m6X_jcd-:-NMAskVu4iU6m6X_jcd0:b-7v1lng) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L809)):*
> ```
> 
> # Querying the row in the local region succeeds.
> query III retry
> ```

Remind me please, why do we need retries on the queries here?
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 899 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAtPRZ4fbqlmno-hPw:-NMAtPRZ4fbqlmno-hPx:b-e4a3rg) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L899)):*
> ```
> 
> ### Now repeat the tests using the row engine.
> subtest enforce_home_region_phase_2_non_vectorized
> ```

We already have `multiregion-9node-3region-3azs-vec-off` config that would do this, should we just add it into the list of `LogicTest` configs at the top?


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Nice work! I left one suggestion for you to consider about an alternative approach for anti-joins, but I think your approach is also fine if you'd rather leave it as-is.

> Reviewed 23 of 23 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-:-NMBGndi9kuBvHVtAUDE:b7mvupa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @msirek)

___
*[`pkg/sql/opt/ops/relational.opt` line 509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMBCbPeDsG7tSbU3ysF:-NMBCbPeDsG7tSbU3ysG:bxyf947) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/opt/ops/relational.opt#L509)):*
> ```
>     # that read into remote regions, though the lookups are defined in
>     # LookupExpr, not RemoteLookupExpr.
>     RemoteOnlyLookups bool
> ```

An alternative approach that wouldn't require execution support, is that when `enforce_home_region` is true, we could replace the remote-only anti join with an "error if rows" node. Not sure if that's better, but it would allow you to remove this boolean and all the execution code related to anti lookup joins.
___
*[`pkg/sql/opt/ops/relational.opt` line 980 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMBBo6C2gHoJTH-ExqQ:-NMBBo6C2gHoJTH-ExqR:btqhw42) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/opt/ops/relational.opt#L980)):*
> ```
>     # search which should error out if it needs to read rows in a remote
>     # region to answer the query.
>     EnforceHomeRegion bool
> ```

Do we need this? Can't we just check the session setting in the execbuilder?


<!-- Sent from Reviewable.io -->

Commit:
physicalplan: pass ctx in more execution code paths

This commit passes the `context.Context` in all code paths necessary to
make it accessible to `PhysicalPlan.MergeResultStreams` and adds a `Ctx`
method to `exec.Factory` to make the context accessible in more locations.

Release note: None

msirek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95484)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @benbardin, @jayshrivastava, @rytaft, and @yuzefovich)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAAbX1BQzfSU9178gs):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we tend to prefix the commit message with a package (or packages) that are mostly affected by the commit, and I don't think we have `exec` package, perhaps `s/exec/sql/`? Also consider changing the commit title to something more understandable.
</blockquote></details>

Modified according to both suggestions
___
*[`pkg/sql/distsql_physical_planner.go` line 3968 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAtxiPDWJHmaPCy7NP:-NMFYT5MAQJYAA44Y9jA:bkka3a8) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/distsql_physical_planner.go#L3968)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: consider adding a quick comment for why we use `1` here.
</blockquote></details>

Added:
```
    // When inputIdx of the SerialUnorderedSynchronizer is incremented to 1,
    // error out.
```
___
*[`pkg/sql/union.go` line 100 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAtpTwCwknpDjP3den:-NMFZFfx29eVB0CpUbKY:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/union.go#L100)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: perhaps mention that `left` input is only accessing local region whereas `right` is all about the remote regions.
</blockquote></details>

Done
___
*[`pkg/sql/execinfrapb/data.proto` line 147 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMALGM63rgWLIP5-RX5:-NMFZX0aBP9f5Qfh2jbJ:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/execinfrapb/data.proto#L147)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: should we explicitly mention that this error must be set if the upper bound is non-zero?
</blockquote></details>

Done
___
*[`pkg/sql/opt/ops/relational.opt` line 509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMBCbPeDsG7tSbU3ysF:-NMF_V4B07-QuaYZuEku:bamb3zf) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/opt/ops/relational.opt#L509)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

An alternative approach that wouldn't require execution support, is that when `enforce_home_region` is true, we could replace the remote-only anti join with an "error if rows" node. Not sure if that's better, but it would allow you to remove this boolean and all the execution code related to anti lookup joins.
</blockquote></details>

That's a good approach, though currently the flag `Memo.enforceHomeRegion` is set to `evalCtx.SessionData().EnforceHomeRegion` and does not check `evalCtx.Planner.IsANSIDML()`.  So if we build an "error if rows" node, we'd have to change the `Memo.enforceHomeRegion` value to equal `evalCtx.SessionData().EnforceHomeRegion && evalCtx.Planner.IsANSIDML()` to make sure that non-DML statements don't error out.  However, the value of `evalCtx.SessionData().EnforceHomeRegion` also affects costing, so if we made this change, we might cache a plan using the modified costs, which could then be used for a query in a session that doesn't have the session setting turned on.  So, I guess keeping the current approach might be simpler.
___
*[`pkg/sql/opt/ops/relational.opt` line 980 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMBBo6C2gHoJTH-ExqQ:-NMFdhb-7dwmUC7NtYRk:brlf3dc) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/opt/ops/relational.opt#L980)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Do we need this? Can't we just check the session setting in the execbuilder?
</blockquote></details>

Removed it
___
*[`pkg/sql/opt/xform/rules/join.opt` line 622 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAoud9C1BRwrCEOXrb:-NMFeqbW9AMoDBitzHlR:bqgwj32) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/opt/xform/rules/join.opt#L622)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: consider mentioning in the comment to this rule that handling of locality-optimized anti joins requires additional execution support, so this rule differs from other types of locality-optimized lookup joins.
</blockquote></details>

Added:
```
# Note: we also apply a similar optimization for other lookup joins; see
# GenerateLocalityOptimizedLookupJoin. The `enforce_home_region` setting
# behavior of erroring out before an attempt to access remote rows is made
# is handled by marking the second anti-join as remote-only, which is
# propagated to structures in the execution engine to trigger the error.
```
___
*[`pkg/sql/physicalplan/physical_plan.go` line 173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMApGB529TBGvL0O1ng:-NMGSxhP71cpbEfC-9HH:bswl752) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/physicalplan/physical_plan.go#L173)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Storing the context here seems like an anti-pattern. We should instead pass `ctx` as an argument to `MergeResultStreams`. The plumbing there shouldn't be too bad.
</blockquote></details>

Added the plumbing in a 2nd commit, plus a way to get the `ctx` via `exec.Factory` (which may be useful in the future).



_Code quote:_
```Go
type SerialStreamErrorSpec struct {
```
___
*[`pkg/sql/rowexec/joinreader.go` line 344 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAq7iy4T2UCg2czxEN:-NMGTYClBUhFswVAw5xV:b-s0kbth) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/rowexec/joinreader.go#L344)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

`EnforceHomeRegion` is part of local-only session data, meaning that it is not propagated to the remote nodes (if the plan is distributed). This shouldn't be an issue since plans with locality-optimized search full run on the gateway (i.e. they are local), but perhaps one day we'd ship it to a remote node to fully run. Should we make some assertions in the code about the plan being fully local? Probably it's not worth the effort, but I thought I'd mention this. WDYT?
</blockquote></details>

Good question. In general locality-optimized search and join is broken if we don't plan it fully local, which would be a problem even if `EnforceHomeRegion` were not set. But I could imagine in the future we may want to support plans with portions which are locality-optimized. Maybe at that time we could move the flag into the set of flags which are propagated to remote nodes. Or, are flags not allowed to be moved? Anyway, it doesn't seem like something which needs to be handled now, unless you can think of a problem case.
___
*[`pkg/sql/rowexec/joinreader.go` line 1026 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAquxS7OSsVTxy_1tw:-NMGbV1S-YyHjgd7B2Hi:b-v08bbs) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/rowexec/joinreader.go#L1026)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: we could extract this out to be a global singleton error that is also reused with the strategy.
</blockquote></details>

OK, I created a singleton error. I didn't understand the comment regarding the strategy.



_Code quote:_
```Go
func (jr *joinReader) performLookup() (joinReaderState, *execinfrapb.ProducerMetadata) {
```
___
*[`pkg/sql/rowflow/input_sync.go` line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMArHbz4QBexPmQmO4l:-NMGfEYGEnsBII5dT6nl:b-du07h8) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/rowflow/input_sync.go#L391)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Why is this a push function rather than an explicit error itself that would be returned on `serialUnorderedSynchronizer.Next` call?
</blockquote></details>

Oh, yeah, I think it would be better to just make a `ProducerMetadata` and return it. Updated as suggested.
___
*[`pkg/sql/rowflow/input_sync.go` line 409 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMArjzc18NSQLe-khha:-NMGgIz903W401NRIKEW:b-pitpl6) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/rowflow/input_sync.go#L409)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: consider adding some commentary here why we don't want to error out with `draining` and `drainBuffered` states.
</blockquote></details>

Added:
```
    // Do not return the error in the `draining` or `drainBuffered` states
    // because we don't want to indicate that a drain operation has
    // failed, and it is only necessary to return the error once.
```
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 13 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAs74q6JY_dPoNQgxt:-NMGhA_63w3owK0eRI7d:brlf3dc) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L13)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

The auto stats collection should already be disabled (it is done in `logic.go:1675`). What prompted you to add this?
</blockquote></details>

Removed it
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 784 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAs_2b9HqHWlzL5U_q:-NMGh5AM2Ih8OeNXk8C_:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L784)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: consider linking the issue number that talks about phase 2.
</blockquote></details>

Done
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 809 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAskVu4iU6m6X_jcd-:-NMGi5_F2QycP562C25T:b2a8e3w) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L809)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Remind me please, why do we need retries on the queries here?
</blockquote></details>

It's due to Zone configs not becoming available in `SystemConfig` between table creation time and query execution time, so we are sometimes missing multiregion properties leading to sporadic mismatches.  See #87391
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 899 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAtPRZ4fbqlmno-hPw:-NMGmgrPBBCINF2JUSOw:b5i3266) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L899)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

We already have `multiregion-9node-3region-3azs-vec-off` config that would do this, should we just add it into the list of `LogicTest` configs at the top?
</blockquote></details>

Thanks for reminding me of this. I tried adding this config, but the problem of zone configs not propagating in a timely manner seems much worse with vectorization off.  Adding some sleep commands alleviated the problem somewhat, but even after adding `sleep 8s` in 3 places, stress tests would still fail.  With vectorize on, stress tests are rock solid with only a single sleep command.  I've added a TODO to try to get the non-vectorized config to work.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> but I'll let @yuzefovich review the last commit

> Reviewed 13 of 13 files at r2, 20 of 20 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-:-NML4TqZE5L8xPreq7B6:b-lzxiud)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @benbardin, @jayshrivastava, @msirek, and @yuzefovich)

___
*[`pkg/sql/opt/ops/relational.opt` line 509 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMBCbPeDsG7tSbU3ysF:-NML2Czh3R03dz-qxPb_:bmss9ni) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/opt/ops/relational.opt#L509)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

That's a good approach, though currently the flag `Memo.enforceHomeRegion` is set to `evalCtx.SessionData().EnforceHomeRegion` and does not check `evalCtx.Planner.IsANSIDML()`.  So if we build an "error if rows" node, we'd have to change the `Memo.enforceHomeRegion` value to equal `evalCtx.SessionData().EnforceHomeRegion && evalCtx.Planner.IsANSIDML()` to make sure that non-DML statements don't error out.  However, the value of `evalCtx.SessionData().EnforceHomeRegion` also affects costing, so if we made this change, we might cache a plan using the modified costs, which could then be used for a query in a session that doesn't have the session setting turned on.  So, I guess keeping the current approach might be simpler.
</blockquote></details>

Ok! Thanks for considering it.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NML3M8hANtg0ODFC6_3:-NML3M8hANtg0ODFC6_4:bi97qna) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1e62db722a53a664db6b7f6d44cf2a1fa779a8/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L3)):*
> ```
> # tenant-cluster-setting-override-opt: allow-multi-region-abstractions-for-secondary-tenants
> # LogicTest: multiregion-9node-3region-3azs !metamorphic-batch-sizes
> # TODO(msirek): Make this test work with the multiregion-9node-3region-3azs-vec-off config
> ```

Any idea why it's not working now?
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NML3G9F0p68vjQU0CwE:-NML3G9G9__6ZIXVln0V:b5jr9ra) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1e62db722a53a664db6b7f6d44cf2a1fa779a8/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L200)):*
> ```
> 
> # Give zone configs time to propagate.
> # sleep 8s
> ```

Since you commented this out, should it be removed?
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 643 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NML3_U41dH6dmStjNOy:-NML3_U41dH6dmStjNOz:bvht12v) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1e62db722a53a664db6b7f6d44cf2a1fa779a8/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L643)):*
> ```
> 
> # Give zone configs time to propagate.
> sleep 5s
> ```

should we use a retry instead?


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95484)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @benbardin, @jayshrivastava, @rytaft, and @yuzefovich)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NML3M8hANtg0ODFC6_3:-NMPJU_YEREMj-Y2KQ6i:b-gh9dpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1e62db722a53a664db6b7f6d44cf2a1fa779a8/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L3)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Any idea why it's not working now?
</blockquote></details>

My guess is that the row engine, being in general slower, is not propagating zone configs in a timely manner. I don't think this is anything new. Other tests like `regional_by_row_query_behavior` have not been using this config either, so the issue is likely pre-existing and is being exposed when we try this config.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 200 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NML3G9F0p68vjQU0CwE:-NMPKJxrBlNqUM916hAw:brlf3dc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1e62db722a53a664db6b7f6d44cf2a1fa779a8/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L200)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Since you commented this out, should it be removed?
</blockquote></details>

Removed it



_Code quote:_
```
sleep 8s
```
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 643 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NML3_U41dH6dmStjNOy:-NMPKPd34VaVjwMxZHXy:b-uvb7e8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1e62db722a53a664db6b7f6d44cf2a1fa779a8/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L643)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

should we use a retry instead?
</blockquote></details>

The test points which fail are issued via `statement error` commands.  I don't think the `retry` directive works with those.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 1 of 1 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95484)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @benbardin, @jayshrivastava, @msirek, and @yuzefovich)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NML3M8hANtg0ODFC6_3:-NMPhBeV-IgmfekWsoRv:bhmn3z9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1e62db722a53a664db6b7f6d44cf2a1fa779a8/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L3)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

My guess is that the row engine, being in general slower, is not propagating zone configs in a timely manner. I don't think this is anything new. Other tests like `regional_by_row_query_behavior` have not been using this config either, so the issue is likely pre-existing and is being exposed when we try this config.
</blockquote></details>

I see. Can you add a bit of that explanation in your TODO?
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 643 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NML3_U41dH6dmStjNOy:-NMPhIexB5wD5jqzVJv-:bivmxn9) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1e62db722a53a664db6b7f6d44cf2a1fa779a8/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L643)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

The test points which fail are issued via `statement error` commands.  I don't think the `retry` directive works with those.
</blockquote></details>

Can we fix the retry directive so it does?


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95484)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @benbardin, @jayshrivastava, @rytaft, and @yuzefovich)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NML3M8hANtg0ODFC6_3:-NMQjzWrAdD2SG22R7vg:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1e62db722a53a664db6b7f6d44cf2a1fa779a8/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L3)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I see. Can you add a bit of that explanation in your TODO?
</blockquote></details>

Done
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 643 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NML3_U41dH6dmStjNOy:-NMQnI5lE9RS5casrTJd:bllf5uw) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1e62db722a53a664db6b7f6d44cf2a1fa779a8/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L643)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Can we fix the retry directive so it does?
</blockquote></details>

Opened #95668 for this, and also to investigate if there is a zone config propagation bug.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 22 of 22 files at r6, 22 of 22 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95484)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @benbardin, @jayshrivastava, and @yuzefovich)

___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 643 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NML3_U41dH6dmStjNOy:-NMRYi2m1qIiWPPRBT3q:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1e62db722a53a664db6b7f6d44cf2a1fa779a8/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L643)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Opened #95668 for this, and also to investigate if there is a zone config propagation bug.
</blockquote></details>

Thanks!


<!-- Sent from Reviewable.io -->

benbardin reviewed (approved):
Approved for `backupccl` <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-:-NMUPyAy3F60LsggNzRW:bdnoewh)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @jayshrivastava and @yuzefovich)



<!-- Sent from Reviewable.io -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 11 of 13 files at r2, 22 of 22 files at r6, 22 of 22 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-:-NMUx9_T6fZP71A6Z3Od:bnfp4nl)* status: :shipit: complete! 2 of 0 LGTMs obtained (and 1 stale) (waiting on @jayshrivastava and @msirek)

___
*[`pkg/sql/plan.go` line 45 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMUwPIADU5oY6EBziaB:-NMUwPIADU5oY6EBziaC:b-mighwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/849a33e04be7dc8552184fb5aca333d630317c43/pkg/sql/plan.go#L45)):*
> ```Go
> 
> // Ctx() gives convenient access to the runParam's `ctx`.
> func (r *runParams) Ctx() context.Context {
> ```

nit: I'd not introduce this method - it seems simple enough to do `params.ctx` rather than `params.Ctx()`, and we already have hundreds usages of the former.
___
*[`pkg/sql/opt/memo/expr.go` line 299 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMUs7P18fnbcoIeh7Eo:-NMUs7P27htNN8QxdxKl:b-y2e2zg) ([raw file](https://github.com/cockroachdb/cockroach/blob/beb93b60440e26e5de91c377359b1a5ad6d1a319/pkg/sql/opt/memo/expr.go#L299)):*
> ```Go
> 		target = distributionProvidedPhysical.Distribution
> 		source = inputDistributionProvidedPhysical.Distribution
> 		return target, source, true
> ```

This seems like a bug fix. Perhaps we're missing some unit tests? Or do you think that this code wasn't really utilized until this PR and this PR introduces tests that caught this bug?
___
*[`pkg/sql/physicalplan/physical_plan.go` line 173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMApGB529TBGvL0O1ng:-NMUwiII4T_3RuZnZcnM:bwwocn3) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/physicalplan/physical_plan.go#L173)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Added the plumbing in a 2nd commit, plus a way to get the `ctx` via `exec.Factory` (which may be useful in the future).
</blockquote></details>

Thanks! I thought it'd be less plumbing though :)
___
*[`pkg/sql/rowexec/joinreader.go` line 344 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAq7iy4T2UCg2czxEN:-NMUsqn_4kuP2RMLhAqB:b-2q3pbm) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/rowexec/joinreader.go#L344)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Good question. In general locality-optimized search and join is broken if we don't plan it fully local, which would be a problem even if `EnforceHomeRegion` were not set. But I could imagine in the future we may want to support plans with portions which are locality-optimized. Maybe at that time we could move the flag into the set of flags which are propagated to remote nodes. Or, are flags not allowed to be moved? Anyway, it doesn't seem like something which needs to be handled now, unless you can think of a problem case.
</blockquote></details>

Yeah, we'll be able to move the flag if needed easily. I agree, let's not spend more time thinking about this.
___
*[`pkg/sql/rowexec/joinreader.go` line 1026 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAquxS7OSsVTxy_1tw:-NMUtJsT5v-H0h1_9YxD:bbh07aj) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/rowexec/joinreader.go#L1026)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

OK, I created a singleton error. I didn't understand the comment regarding the strategy.
</blockquote></details>

I meant that we can replace the errors constructed in two places in `joinreader_strategies.go` with the global singleton introduced in `joinreader.go`.
___
*[`pkg/sql/rowflow/input_sync.go` line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMArHbz4QBexPmQmO4l:-NMUtqr59aE1jWeXF3Lp:bizya54) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/rowflow/input_sync.go#L391)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Oh, yeah, I think it would be better to just make a `ProducerMetadata` and return it. Updated as suggested.
</blockquote></details>

nit: to me it'd seem a little cleaner if this function returned an error and then the synchronizer would wrap it in `ProducerMeta`. My rationale is that this is an "error func", so based on the name I expect it to return an error. I don't have a strong feeling though, so please feel free to ignore this comment.
___
*[`pkg/sql/rowflow/row_based_flow.go` line 335 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMUuNyp4xho8LsHPyHB:-NMUuNyp4xho8LsHPyHC:b-8ie9zv) ([raw file](https://github.com/cockroachdb/cockroach/blob/beb93b60440e26e5de91c377359b1a5ad6d1a319/pkg/sql/rowflow/row_based_flow.go#L335)):*
> ```Go
> 					exceedsSrcIndexExclusiveUpperBoundError := is.EnforceHomeRegionError.ErrorDetail(ctx)
> 					returnErrorFunc = func() *execinfrapb.ProducerMetadata {
> 						return &execinfrapb.ProducerMetadata{Err: exceedsSrcIndexExclusiveUpperBoundError}
> ```

nit: you could use `execinfrapb.GetProducerMeta` to get an object from the sync pool (it'll be automatically put back into the pool in `DistSQLReceiver.pushMeta`).
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 809 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAskVu4iU6m6X_jcd-:-NMUvI0XEp2exvtIZbtW:b-k2awqz) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L809)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

It's due to Zone configs not becoming available in `SystemConfig` between table creation time and query execution time, so we are sometimes missing multiregion properties leading to sporadic mismatches.  See #87391
</blockquote></details>

Ah, ok, thanks.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 899 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAtPRZ4fbqlmno-hPw:-NMUvUvk9Fedmz9BPQKG:b9xvvbl) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L899)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Thanks for reminding me of this. I tried adding this config, but the problem of zone configs not propagating in a timely manner seems much worse with vectorization off.  Adding some sleep commands alleviated the problem somewhat, but even after adding `sleep 8s` in 3 places, stress tests would still fail.  With vectorize on, stress tests are rock solid with only a single sleep command.  I've added a TODO to try to get the non-vectorized config to work.
</blockquote></details>

Ack, thanks for trying.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NML3M8hANtg0ODFC6_3:-NMUupci3eNvYnJ8lUg7:b-s6hyfn) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1e62db722a53a664db6b7f6d44cf2a1fa779a8/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L3)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Done
</blockquote></details>

Hm, I'm surprised by this - propagation of the cluster settings' updates is independent of the execution engine (the propagation is done at a lower level). I'm ok leaving this as a TODO, but I'm still puzzled why we'd see a difference between these two logic test configs.
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 204 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMUwxSU2KsHu6fIqeeh:-NMUwxSU2KsHu6fIqeei:b-25cxmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/beb93b60440e26e5de91c377359b1a5ad6d1a319/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L204)):*
> ```
> messages_global
> 
> # Give zone configs time to propagate.
> ```

nit: this should be removed in this first commit.


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95484)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @jayshrivastava, @rytaft, and @yuzefovich)

___
*[`pkg/sql/opt/memo/expr.go` line 299 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMUs7P18fnbcoIeh7Eo:-NMV80Qv7YAiocbImWQI:b-8wc7l7) ([raw file](https://github.com/cockroachdb/cockroach/blob/beb93b60440e26e5de91c377359b1a5ad6d1a319/pkg/sql/opt/memo/expr.go#L299)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

This seems like a bug fix. Perhaps we're missing some unit tests? Or do you think that this code wasn't really utilized until this PR and this PR introduces tests that caught this bug?
</blockquote></details>

I've added a comment to the test which caught this bug:
```
# Locality-optimized semijoin reading into a remote region fails.
# Prior to fixing a bug in `DistributeExpr.GetDistributions`, this query would
# return a different error message.
statement error pq: Query has no home region\. Try using a lower LIMIT value or running the query from a different region\.
SELECT message FROM (SELECT * FROM messages_rbt ORDER BY account_id LIMIT 2) rbt WHERE account_id IN
                     (SELECT account_id FROM messages_rbr rbr) ORDER BY 1 LIMIT 2;
```

I've also added unit test `TestGetDistributions`.
___
*[`pkg/sql/rowexec/joinreader.go` line 344 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAq7iy4T2UCg2czxEN:-NMVY-632Rcn1eOI9guh:b2nhysk) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/rowexec/joinreader.go#L344)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Yeah, we'll be able to move the flag if needed easily. I agree, let's not spend more time thinking about this.
</blockquote></details>

ack
___
*[`pkg/sql/rowexec/joinreader.go` line 1026 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAquxS7OSsVTxy_1tw:-NMVYjvYALTENu6_1RTi:b9wveop) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/rowexec/joinreader.go#L1026)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I meant that we can replace the errors constructed in two places in `joinreader_strategies.go` with the global singleton introduced in `joinreader.go`.
</blockquote></details>

OK, made the change.



_Code quote:_
```Go
// performLookup reads the next batch of index rows.
```
___
*[`pkg/sql/rowflow/input_sync.go` line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMArHbz4QBexPmQmO4l:-NMVZtiA9RZA2UwC3OzE:baebqfz) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/sql/rowflow/input_sync.go#L391)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: to me it'd seem a little cleaner if this function returned an error and then the synchronizer would wrap it in `ProducerMeta`. My rationale is that this is an "error func", so based on the name I expect it to return an error. I don't have a strong feeling though, so please feel free to ignore this comment.
</blockquote></details>

Updated as suggested



_Code quote:_
```Go
exceedsSrcIndexExclusiveUpperBoundErrorFunc func() *execinfrapb.ProducerMetadata
```
___
*[`pkg/sql/rowflow/row_based_flow.go` line 335 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMUuNyp4xho8LsHPyHB:-NMV_JY48N2Kb_lWL47M:bkiaku8) ([raw file](https://github.com/cockroachdb/cockroach/blob/beb93b60440e26e5de91c377359b1a5ad6d1a319/pkg/sql/rowflow/row_based_flow.go#L335)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: you could use `execinfrapb.GetProducerMeta` to get an object from the sync pool (it'll be automatically put back into the pool in `DistSQLReceiver.pushMeta`).
</blockquote></details>

Made the change in `serialUnorderedSynchronizer.Next()`
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 809 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAskVu4iU6m6X_jcd-:-NMVedkN2mWw1lqFy4ge:bb74ce2) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L809)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Ah, ok, thanks.
</blockquote></details>

Sure
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 899 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMAtPRZ4fbqlmno-hPw:-NMVejzD3ZiNtcIvFm3n:bb74ce2) ([raw file](https://github.com/cockroachdb/cockroach/blob/daf6f0d5d141f9b9540c41226226698ac244a6ea/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L899)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Ack, thanks for trying.
</blockquote></details>

Sure
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 3 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NML3M8hANtg0ODFC6_3:-NMV_PcSAh2T1wbHgQRR:b2nhysk) ([raw file](https://github.com/cockroachdb/cockroach/blob/5d1e62db722a53a664db6b7f6d44cf2a1fa779a8/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L3)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, I'm surprised by this - propagation of the cluster settings' updates is independent of the execution engine (the propagation is done at a lower level). I'm ok leaving this as a TODO, but I'm still puzzled why we'd see a difference between these two logic test configs.
</blockquote></details>

ack
___
*[`pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error` line 204 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMUwxSU2KsHu6fIqeeh:-NMVeTahB0qshQpj4HCn:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/beb93b60440e26e5de91c377359b1a5ad6d1a319/pkg/ccl/logictestccl/testdata/logic_test/multi_region_remote_access_error#L204)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: this should be removed in this first commit.
</blockquote></details>

Done
___
*[`pkg/sql/plan.go` line 45 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-NMUwPIADU5oY6EBziaB:-NMV7EIb3-UJuLomGGFB:bbeg432) ([raw file](https://github.com/cockroachdb/cockroach/blob/849a33e04be7dc8552184fb5aca333d630317c43/pkg/sql/plan.go#L45)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: I'd not introduce this method - it seems simple enough to do `params.ctx` rather than `params.Ctx()`, and we already have hundreds usages of the former.
</blockquote></details>

OK, removed it.



_Code quote:_
```Go
func (r *runParams) Ctx() context.Context {
```


<!-- Sent from Reviewable.io -->

msirek reviewed (commented):
TFTRs!
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/95484#-:-NMWDNDB708HUGzFEhJK:bg1822s)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 3 stale) (waiting on @jayshrivastava, @rytaft, and @yuzefovich)



<!-- Sent from Reviewable.io -->
